hedera-node/hedera-app/src/main/java/com/hedera/node/app/state/recordcache/DeduplicationCacheImpl.java

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright (C) 2023 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hedera.node.app.state.recordcache;
+
+import static com.hedera.node.app.spi.HapiUtils.TIMESTAMP_COMPARATOR;
+import static com.hedera.node.app.spi.HapiUtils.asTimestamp;
+import static com.hedera.node.app.spi.HapiUtils.minus;
+import static java.util.Objects.requireNonNull;
+
+import com.hedera.hapi.node.base.TransactionID;
+import com.hedera.node.app.service.mono.context.properties.GlobalDynamicProperties;
+import com.hedera.node.app.state.DeduplicationCache;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.time.Instant;
+import java.util.Set;
+import java.util.TreeSet;
+import javax.inject.Inject;
+
+/** An implementation of {@link DeduplicationCache}. */
+public final class DeduplicationCacheImpl implements DeduplicationCache {
+    /**
+     * The {@link TransactionID}s that this node has already submitted to the platform, sorted by transaction start
+     * time, such that earlier start times come first. We guard this data structure within a synchronized block.
+     */
+    private final Set<TransactionID> submittedTxns = new TreeSet<>((t1, t2) ->
+            TIMESTAMP_COMPARATOR.compare(t1.transactionValidStartOrThrow(), t2.transactionValidStartOrThrow()));
+
+    /** Used for looking up the max transaction duration window. To be replaced by some new config object */
+    private final GlobalDynamicProperties props;
+
+    /** Constructs a new {@link DeduplicationCacheImpl}. */
+    @Inject
+    public DeduplicationCacheImpl(@NonNull final GlobalDynamicProperties props) {
+        this.props = requireNonNull(props);
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * <p>This method is not called at a super high rate, so synchronizing here is perfectly fine.
```

Comment:
Isn't this also being used in `preHandle()`, so at 10k TPS called 10k times per second?

Reply:
Yes you are right. The comment is not right, and it is possibly then a source of contention. Using a better threadsafe data structure might be required when we get to performance profiling.

Reply:
I removed the comment for now.

Reply:
Quick test with 1 million items is that ConcurrentSkipListSet is 3 microseconds per action faster than synchronized keyword.